core/tools/size_firmware_elf.py

```diff
@@ -0,0 +1,238 @@
+"""
+
+from __future__ import annotations
+
+import re
+import shutil
+import subprocess
+from fnmatch import fnmatch
+from pathlib import Path
+
+import click
+
+from bin_size import BinarySize, DataRow, StatisticsPlugin, show_binaries_diff
+
+HERE = Path(__file__).parent
+CORE_DIR = HERE.parent
+
+BUILD_CMD = f"(cd {CORE_DIR} && make build_firmware)"
+FIRMWARE_ELF = CORE_DIR / "build/firmware/firmware.elf"
+
+
+def _build_binary(extra_suffix: str | None = None) -> None:
+    print(f"building the binary... `{BUILD_CMD}`")
+    build_result = subprocess.run(
+        BUILD_CMD, stdout=subprocess.PIPE, text=True, shell=True
+    )
+    if build_result.returncode != 0:
+        print("build failed - see output above")
+        exit(1)
+
+    # Optionally copying the binary to a special name, to be used later
+    if extra_suffix is not None:
+        new_path = Path(f"{FIRMWARE_ELF}_{extra_suffix}")
+        shutil.copyfile(FIRMWARE_ELF, new_path)
+        print(f"Binary copied as `{new_path}`")
+
+
+def _handle_apps_statistics(BS: BinarySize, output_file: str) -> None:
+    def apps_categories(row: DataRow) -> str | None:
+        pattern = r"^src/apps/(\w+)/"  # dir name after apps/
+        match = re.search(pattern, row.module_name)
+        if not match:
+            return None
+        else:
+            return match.group(1)
+
+    StatisticsPlugin(BS, apps_categories).show(output_file or None)
+
+
+#
+# CLI PART
+#
+
+
+@click.group()
+def cli() -> None:
+    pass
+
+
+@cli.command()
+@click.option(
+    "-f", "--custom-file", help="Analyze custom file instead of the default one"
+)
+@click.option("-o", "--output-file", help="Dump results to file instead of stdout")
+@click.option("-a", "--app", help="App which to analyze - e.g. `ethereum`")
+@click.option(
+    "-l",
+    "--language",
+    type=click.Choice(("mpy", "Rust", "C")),
+    help="Language choice - e.g. `Rust`",
+)
+@click.option(
+    "-g",
+    "--grep",
+    help="Custom string to filter the row with, case-insensitive - e.g. `bitcoin`",
+)
+@click.option(
+    "-mn",
+    "--module-name",
+    help="Check only specific file/module. Supports shell-style wildcards - e.g. `*/networks.py`",
+)
+@click.option(
+    "-fn", "--func-name", help="Check only specific function - e.g. `get_tx_keys`"
+)
+@click.option("-b", "--build", is_flag=True, help="Perform build")
+@click.option(
+    "-as", "--all-sections", is_flag=True, help="Get all sections, not just flash"
+)
+@click.option(
+    "-s", "--apps-statistics", is_flag=True, help="Overall statistics for all apps"
+)
+@click.option(
+    "-ad",
+    "--add-definitions",
+    is_flag=True,
+    help="Get line definitions for all functions",
+)
+@click.option(
+    "-na", "--no-aggregation", is_flag=True, help="Do not aggregate symbols together"
+)
+@click.option("-ns", "--no-sort", is_flag=True, help="Do not sort by size")
+@click.option("-np", "--no-processing", is_flag=True, help="See just raw data")
```

Comment:
short option should always be a single letter, so `-D` not `-ad`
(the point is, you can then group multiple short options together: instead of `-a -b -c -C -d`, a single `-abcCd` is a valid way to specify. using more than one letter breaks this functionality.
if you want two-letter shortcuts, it is perfectly valid to specify `"-D", "--ad", "--add-definitions"`)